pkg/kubectl/cmd/util/factory_object_mapping.go

```diff
@@ -398,10 +401,95 @@ func (f *ring1Factory) AttachablePodForObject(object runtime.Object, timeout tim
 	}
 
 	sortBy := func(pods []*v1.Pod) sort.Interface { return sort.Reverse(controller.ActivePods(pods)) }
-	pod, _, err := GetFirstPod(clientset.Core(), namespace, selector, timeout, sortBy)
+	pod, _, err := GetFirstPod(clientset.Core(), namespace, selector, timeout, true, sortBy)
 	return pod, err
 }
 
+func (f *ring1Factory) MostAccuratePodTemplateForObject(object runtime.Object) (*api.PodTemplateSpec, error) {
+	sortBy := func(pods []*v1.Pod) sort.Interface { return controller.ActivePods(pods) }
+	clientset, err := f.clientAccessFactory.ClientSetForVersion(nil)
+	if err != nil {
+		return nil, err
+	}
+	var podSpec *api.PodTemplateSpec
+	var pod *api.Pod
+	var selector labels.Selector
+	var namespace string
+	switch t := object.(type) {
+	case *extensions.ReplicaSet:
+		namespace = t.Namespace
+		selector = labels.SelectorFromSet(t.Spec.Selector.MatchLabels)
+		if err != nil {
+			return &t.Spec.Template, fmt.Errorf("invalid label selector: %v", err)
+		}
+	case *api.ReplicationController:
+		namespace = t.Namespace
+		selector = labels.SelectorFromSet(t.Spec.Selector)
+		if err != nil {
+			return t.Spec.Template, fmt.Errorf("invalid label selector: %v", err)
+		}
+	case *apps.StatefulSet:
+		namespace = t.Namespace
+		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
+		if err != nil {
+			return &t.Spec.Template, fmt.Errorf("invalid label selector: %v", err)
+		}
+	case *extensions.Deployment:
+		namespace = t.Namespace
+		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
+		if err != nil {
+			return &t.Spec.Template, fmt.Errorf("invalid label selector: %v", err)
+		}
+		options := metav1.ListOptions{LabelSelector: selector.String()}
+		var allRs []*extensions_v1beta1.ReplicaSet
+		rsList, err := clientset.Extensions().ReplicaSets(namespace).List(options)
+		if err != nil {
+			return &t.Spec.Template, err
+		}
+		for _, rs := range rsList.Items {
+			rs_external := &extensions_v1beta1.ReplicaSet{}
+			k8s_extentions_v1beta1.Convert_extensions_ReplicaSet_To_v1beta1_ReplicaSet(&rs, rs_external, nil)
+			allRs = append(allRs, rs_external)
+		}
+		d := &extensions_v1beta1.Deployment{}
+		api.Scheme.Convert(t, d, nil)
+		rs, err := deploymentutil.FindNewReplicaSet(d, allRs)
+		if err != nil {
+			return &t.Spec.Template, err
+		}
+		selector, err = metav1.LabelSelectorAsSelector(rs.Spec.Selector)
+		if err != nil {
+			return &t.Spec.Template, fmt.Errorf("invalid label selector: %v", err)
+		}
+	case *batch.Job:
+		namespace = t.Namespace
+		selector, err = metav1.LabelSelectorAsSelector(t.Spec.Selector)
+		if err != nil {
+			return &t.Spec.Template, fmt.Errorf("invalid label selector: %v", err)
+		}
+	case *api.Pod:
+		podSpec = &api.PodTemplateSpec{
+			ObjectMeta: t.ObjectMeta,
+			Spec:       t.Spec,
+		}
+		return podSpec, nil
+	default:
```

Comment:
Additionally, I don't like this switch, it's a 3rd aside from almost identical in `AttachablePodForObject` and `LogsForObject`. Could you try cleaning this a bit?